### PR TITLE
fix(logspout): fix deadlock when attaching

### LINKED ISSUE #3633: 'deis logs' not catching all app logs from docker containers
docker logs
```
Wed May 06 2015 07:44:45 GMT+0000 (UTC): Webhook detected, rebuilding website started..
Wed May 06 2015 07:45:13 GMT+0000 (UTC): Rebuilding website completed.
Wed May 06 2015 07:47:21 GMT+0000 (UTC): Webhook detected, rebuilding website started..
Wed May 06 2015 07:47:50 GMT+0000 (UTC): Rebuilding website completed.
Wed May 06 2015 08:06:08 GMT+0000 (UTC): Webhook detected, rebuilding website started..
Wed May 06 2015 08:06:37 GMT+0000 (UTC): Rebuilding website completed.
Wed May 06 2015 08:08:58 GMT+0000 (UTC): Webhook detected, rebuilding website started..
Wed May 06 2015 08:09:27 GMT+0000 (UTC): Rebuilding website completed.
Wed May 06 2015 08:25:35 GMT+0000 (UTC): Webhook detected, rebuilding website started..
Wed May 06 2015 08:26:02 GMT+0000 (UTC): Rebuilding website completed.
Wed May 06 2015 08:29:41 GMT+0000 (UTC): Webhook detected, rebuilding website started..
Wed May 06 2015 08:30:09 GMT+0000 (UTC): Rebuilding website completed.
Wed May 06 2015 08:41:58 GMT+0000 (UTC): Webhook detected, rebuilding website started..
Wed May 06 2015 08:42:26 GMT+0000 (UTC): Rebuilding website completed.
Wed May 06 2015 08:48:18 GMT+0000 (UTC): Webhook detected, rebuilding website started..
Wed May 06 2015 08:48:46 GMT+0000 (UTC): Rebuilding website completed.
Wed May 06 2015 09:01:14 GMT+0000 (UTC): Webhook detected, rebuilding website started..
Wed May 06 2015 09:01:43 GMT+0000 (UTC): Rebuilding website completed.
Wed May 06 2015 09:05:35 GMT+0000 (UTC): Webhook detected, rebuilding website started..
Wed May 06 2015 09:06:03 GMT+0000 (UTC): Rebuilding website completed.
Wed May 06 2015 09:08:18 GMT+0000 (UTC): Webhook detected, rebuilding website started..
Wed May 06 2015 09:08:47 GMT+0000 (UTC): Rebuilding website completed.
```

deis client logs
```
2015-05-06T09:54:15UTC vaer[web.1]: Wed May 06 2015 09:54:15 GMT+0000 (UTC): Building website to /public..
2015-05-06T09:54:46UTC vaer[web.1]: Wed May 06 2015 09:54:46 GMT+0000 (UTC): Building website completed.
2015-05-06T09:58:47UTC vaer[web.1]: Wed May 06 2015 09:58:47 GMT+0000 (UTC): Rebuilding website completed.
2015-05-06T10:07:39UTC vaer[web.1]: Wed May 06 2015 10:07:39 GMT+0000 (UTC): Webhook detected, rebuilding website started..
2015-05-06T10:08:08UTC vaer[web.1]: Wed May 06 2015 10:08:08 GMT+0000 (UTC): Rebuilding website completed.
2015-05-06T10:08:56UTC vaer[web.1]: Wed May 06 2015 10:08:56 GMT+0000 (UTC): Rebuilding website completed.
2015-05-06T10:18:19UTC vaer[web.1]: Wed May 06 2015 10:18:19 GMT+0000 (UTC): Rebuilding website completed.
2015-05-06T10:19:08UTC vaer[web.1]: Wed May 06 2015 10:19:08 GMT+0000 (UTC): Rebuilding website completed.
2015-05-06T10:28:01UTC vaer[web.1]: Wed May 06 2015 10:28:01 GMT+0000 (UTC): Webhook detected, rebuilding website started..
2015-05-06T10:28:32UTC vaer[web.1]: Wed May 06 2015 10:28:32 GMT+0000 (UTC): Rebuilding website completed.
2015-05-06T10:28:51UTC vaer[web.1]: Wed May 06 2015 10:28:51 GMT+0000 (UTC): Webhook detected, rebuilding website started..
2015-05-06T10:29:21UTC vaer[web.1]: Wed May 06 2015 10:29:21 GMT+0000 (UTC): Rebuilding website completed.
2015-05-06T10:38:15UTC vaer[web.1]: Wed May 06 2015 10:38:15 GMT+0000 (UTC): Webhook detected, rebuilding website started..
2015-05-06T10:38:43UTC vaer[web.1]: Wed May 06 2015 10:38:43 GMT+0000 (UTC): Rebuilding website completed.
2015-05-06T10:39:35UTC vaer[web.1]: Wed May 06 2015 10:39:35 GMT+0000 (UTC): Rebuilding website completed.
2015-05-06T11:53:31UTC deis[api]: hivearts deployed 99f07e0
2015-05-06T11:53:58UTC vaer[web.1]: Listening on port 5000
2015-05-06T11:53:58UTC vaer[web.1]: Wed May 06 2015 11:53:58 GMT+0000 (UTC): Building website to /public..
2015-05-06T11:54:28UTC vaer[web.1]: Wed May 06 2015 11:54:28 GMT+0000 (UTC): Building website completed.
2015-05-06T11:58:37UTC vaer[web.1]: Wed May 06 2015 11:58:37 GMT+0000 (UTC): Rebuilding website completed.
2015-05-06T12:00:46UTC vaer[web.1]: Wed May 06 2015 12:00:46 GMT+0000 (UTC): Rebuilding website completed.
```

As you can see deis not catching all logs from docker container, what should i do?
I am using v1.5.2 with v1.6.0-rc1 router.

### PR DESCRIPTION
**Pull request by krancour**
Fixes #3633 

This isn't guaranteed to be a magic bullet that solves every logging-related issue, but we have uncovered deadlock that can occur when logspout attempts attaching to containers.  At several points in this process, a lock on a mutex is acquired and released.  At one such point, after acquiring a lock, a message is written to an unbuffered channel.  Since writes to an unbuffered channel will block until another goroutine receives the message, and since sometimes there is no one listening, execution can block indefinitely at this point while holding on to the lock that other goroutines are waiting for.  Using a buffered channel seems to alleviate this issue.

Thanks to @aledbf and @technosophos for help in diagnosing the issue.

N.B.: Longer term, we might consider adopting a newer version of logspout, as it's gone through two major refactors since we forked and enhanced it to work with Deis.  I'll track that in a separate issue.

### CODE REVIEW

**Comment by carmstrong** [comment]
Deferring to @technosophos / @kalbasit / @aledbf for review, but I trust your debugging. :+1:

**Comment by kalbasit** [comment]
It looks like a reasonable solution, quick questions though as I am unfamiliar with the logspout code:
- Why only 1? Is the event channel never receive more than one message? It seems to me that it's receive `len(m.attached)` messages [here](https://github.com/krancour/deis/blob/fix-logspout-deadlock/logspout/attacher.go#L104)
- What line are you seeing the deadlock exactly?

I agree with you, the code looks quite unstable, perhaps it's best to update logspout as part of this release. @carmstrong thoughts?

**Comment by carmstrong** [comment]
> perhaps it's best to update logspout as part of this release

I think that's a worth endeavor if we think it may resolve this issue.

**Comment by kalbasit** [comment]
@progrium was this issue resolved with the refactor?

**Comment by technosophos** [comment]
Setting a channel to length 1 means that the sender will not block on send. So it will most likely solve the problem. But it will cause blocking if the reader is not keeping up with the writer. So @kalbasit's suggestion may improve performance.

It's unclear to me whether `m.attached` can change during the processing of that goroutine. `addListener` may have another hidden race condition in there. Since iteration order on a map isn't guaranteed, I'm not exactly sure what would happen on that range loop if something else was modifying the map.

**Comment by carmstrong** [comment]
Have we evaluated the upstream changes? If they appear to address this issue and the map iteration, we should 

I say we merge this as-is and focus on the refactor proposed in #4292.

**Comment by krancour** [comment]
@carmstrong upstream's been completely refactored.  I can't say there are no issues, but the bits of code that were causing deadlock are long gone.  :+1: on merging this as is to get a little relief while we focus on a more strategic overhaul on logspout.

**Comment by kalbasit** [comment]
I'm fine with this plan @krancour but perhaps it would be wise to change the lengh of the channel from 1 to `len(m.attached)` to eliminate a possibility of another deadlock.

**Comment by krancour** [comment]
I'm not sure I understand why channel depth of `len(m.attached)`.  That means every time a channel is established to be notified of (future) attach events, the depth of the channel is dependent on how many containers we're already attached to.  I'm not sure if that relationship makes sense.

**Comment by krancour** [comment]
Although I am not arguing that depth > 1 might be advisable.  I am only questioning whether `len(m.attached)` is just as arbitrary as saying `5` or `10`.

**Comment by kalbasit** [comment]
@krancour it's mostly because of [this](https://github.com/krancour/deis/blob/fix-logspout-deadlock/logspout/attacher.go#L102-L106) which ranges over `m.attached` and send the `events` channel a message. It is arbitrary and yes it might still break further down the line but it's sure better than the fixed one, well at least for the time being until we figure out the right course for #4292

**Comment by krancour** [comment]
@kalbasit thanks.  This code is a bit hard to follow in places.  I agree with you.  I had overlooked that when adding a listener, an `AttachEvent` is sent to that channel for every container we're already attached to.  I agree with your proposed change, but I'm going to amend it so depth is 1 if `len(m.attached)` is 0.  That condition could occur and could create an unbuffered channel if we're no on guard.

**Comment by progrium** [comment]
I'm wondering if this version is too old to contribute back to logspout proper but that would be nice.

**Comment by krancour** [comment]
@progrium this is quite old indeed.  We're tracking a separate issue to bring ourselves more up to date and hopefully do so without having to fork logspout, per your documentation in master.

**Comment by krancour** [comment]
@kalbasit's advice has been applied.  Thanks @kalbasit !

**Comment by mboersma** [comment]
This band-aid LGTM.

**Comment by bacongobbler** [comment]
code LGTM
